hazelcast-hibernate53/src/main/java/com/hazelcast/hibernate/AbstractHazelcastCacheRegionFactory.java

```diff
@@ -0,0 +1,174 @@
+
+    private final CacheKeysFactory cacheKeysFactory;
+    private final ILogger log = Logger.getLogger(getClass());
+
+    private IHazelcastInstanceLoader instanceLoader;
+
+    @SuppressWarnings("unused")
+    public AbstractHazelcastCacheRegionFactory() {
+        this(DefaultCacheKeysFactory.INSTANCE);
+    }
+
+    public AbstractHazelcastCacheRegionFactory(final CacheKeysFactory cacheKeysFactory) {
+        this.cacheKeysFactory = cacheKeysFactory;
+    }
+
+    public AbstractHazelcastCacheRegionFactory(final HazelcastInstance instance) {
+        this.instance = instance;
+
+        cacheKeysFactory = DefaultCacheKeysFactory.INSTANCE;
+    }
+
+    @Override
+    public DomainDataRegion buildDomainDataRegion(final DomainDataRegionConfig regionConfig,
+                                                  final DomainDataRegionBuildingContext buildingContext) {
+        return new HazelcastDomainDataRegionImpl(
+                regionConfig,
+                this,
+                createDomainDataStorageAccess(regionConfig, buildingContext),
+                cacheKeysFactory,
+                buildingContext
+        );
+    }
+
+    public HazelcastInstance getHazelcastInstance() {
+        return instance;
+    }
+
+    @Override
+    protected DomainDataStorageAccess createDomainDataStorageAccess(final DomainDataRegionConfig regionConfig,
+                                                                    final DomainDataRegionBuildingContext buildingContext) {
+        return new HazelcastStorageAccessImpl(
+                createRegionCache(regionConfig.getRegionName(), buildingContext.getSessionFactory(), regionConfig)
+        );
+    }
+
+    @Override
+    protected StorageAccess createQueryResultsRegionStorageAccess(final String regionName,
+                                                                  final SessionFactoryImplementor sessionFactory) {
+        // Note: We don't want to use an ITopic for invalidation because the timestamps cache can take care of outdated
+        // queries
+        final LocalRegionCache regionCache = new LocalRegionCache(this, regionName, instance, null, false);
+        cleanupService.registerCache(regionCache);
+        return new HazelcastStorageAccessImpl(regionCache);
+    }
+
+    protected abstract RegionCache createRegionCache(String unqualifiedRegionName,
+                                                     SessionFactoryImplementor sessionFactory,
+                                                     DomainDataRegionConfig regionConfig);
+
+    @Override
+    protected StorageAccess createTimestampsRegionStorageAccess(final String regionName,
+                                                                final SessionFactoryImplementor sessionFactory) {
+        return new HazelcastStorageAccessImpl(
+                createTimestampsRegionCache(regionName, sessionFactory)
+        );
+    }
+
+    protected abstract RegionCache createTimestampsRegionCache(String regionName,
```

Comment:
More here. Perhaps one of the Hazelcast maintainers can link us to something re: their code style

Reply:
Done.